drivers/usb/dwc3/core.c

```diff
@@ -566,6 +566,11 @@ static int dwc3_phy_setup(struct dwc3 *dwc)
 		ret = dwc3_ulpi_init(dwc);
 		if (ret)
 			return ret;
+		if(dwc->dr_mode == USB_DR_MODE_HOST)
+		{
+                        reg |= DWC3_GUSB2PHYCFG_ULPI_EXT_VBUS_DRV;
```

Comment:
ULPI accesses on dwc3 are timing out. I'll wait for Heikki to come back from his time off and ask him if he's seen this before. Here's current hack/debug patch:

```
diff --git a/drivers/phy/phy-tusb1210.c b/drivers/phy/phy-tusb1210.c
index 4f6d5e71507d..3d6e39611985 100644
--- a/drivers/phy/phy-tusb1210.c
+++ b/drivers/phy/phy-tusb1210.c
@@ -11,6 +11,7 @@
  */
 #include <linux/module.h>
 #include <linux/ulpi/driver.h>
+#include <linux/ulpi/regs.h>
 #include <linux/gpio/consumer.h>
 
 #include "ulpi_phy.h"
@@ -31,6 +32,7 @@ struct tusb1210 {
 static int tusb1210_power_on(struct phy *phy)
 {
 	struct tusb1210 *tusb = phy_get_drvdata(phy);
+	int ret;
 
 	gpiod_set_value_cansleep(tusb->gpio_reset, 1);
 	gpiod_set_value_cansleep(tusb->gpio_cs, 1);
@@ -40,7 +42,14 @@ static int tusb1210_power_on(struct phy *phy)
 		ulpi_write(tusb->ulpi, TUSB1210_VENDOR_SPECIFIC2,
 			   tusb->vendor_specific2);
 
-	return 0;
+	/* TUSB1210 only supports external charge pump */
+	ret = ulpi_read(tusb->ulpi, ULPI_OTG_CTRL);
+	if (ret < 0)
+		return ret;
+
+	ret |= ULPI_OTG_CTRL_DRVVBUS_EXT;
+
+	return ulpi_write(tusb->ulpi, ULPI_OTG_CTRL, ret);
 }
 
 static int tusb1210_power_off(struct phy *phy)
@@ -53,9 +62,43 @@ static int tusb1210_power_off(struct phy *phy)
 	return 0;
 }
 
+static int tusb1210_set_mode(struct phy *phy, enum phy_mode mode)
+{
+	struct tusb1210 *tusb = phy_get_drvdata(phy);
+	int ret;
+
+	printk(KERN_INFO "==> %s %d\n", __func__, __LINE__);
+	ret = ulpi_read(tusb->ulpi, ULPI_OTG_CTRL);
+	if (ret < 0) {
+		printk(KERN_INFO "==> %s %d --> %d\n", __func__, __LINE__, ret);
+		return ret;
+	}
+	printk(KERN_INFO "==> %s %d\n", __func__, __LINE__);
+
+	switch (mode) {
+	case PHY_MODE_USB_HOST:
+		printk(KERN_INFO "==> %s %d\n", __func__, __LINE__);
+		ret |= ULPI_OTG_CTRL_DRVVBUS;
+		break;
+	case PHY_MODE_USB_DEVICE:
+		printk(KERN_INFO "==> %s %d\n", __func__, __LINE__);
+		ret &= ~ULPI_OTG_CTRL_DRVVBUS;
+		break;
+	default:
+		printk(KERN_INFO "==> %s %d\n", __func__, __LINE__);
+		/* nothing */
+		return 0;
+	}
+
+	printk(KERN_INFO "===> %s CTRL %08x\n", __func__, ret);
+
+	return ulpi_write(tusb->ulpi, ULPI_OTG_CTRL, ret);
+}
+
 static const struct phy_ops phy_ops = {
 	.power_on = tusb1210_power_on,
 	.power_off = tusb1210_power_off,
+	.set_mode = tusb1210_set_mode,
 	.owner = THIS_MODULE,
 };
 
@@ -125,7 +168,8 @@ static void tusb1210_remove(struct ulpi *ulpi)
 #define TI_VENDOR_ID 0x0451
 
 static const struct ulpi_device_id tusb1210_ulpi_id[] = {
-	{ TI_VENDOR_ID, 0x1507, },
+	{ TI_VENDOR_ID, 0x1507, },  /* TUSB1210 */
+	{ TI_VENDOR_ID, 0x1508, },  /* TUSB1211 */
 	{ },
 };
 MODULE_DEVICE_TABLE(ulpi, tusb1210_ulpi_id);
diff --git a/drivers/usb/dwc3/core.c b/drivers/usb/dwc3/core.c
index 455d89a1cd6d..9124f67b17b7 100644
--- a/drivers/usb/dwc3/core.c
+++ b/drivers/usb/dwc3/core.c
@@ -150,12 +150,33 @@ static void __dwc3_set_mode(struct work_struct *work)
 
 	switch (dwc->desired_dr_role) {
 	case DWC3_GCTL_PRTCAP_HOST:
+		printk(KERN_INFO "==> %s %d\n", __func__, __LINE__);
+		if (dwc->usb2_phy) {
+		printk(KERN_INFO "==> %s %d\n", __func__, __LINE__);
+			otg_set_vbus(dwc->usb2_phy->otg, true);
+		}
+		printk(KERN_INFO "==> %s %d\n", __func__, __LINE__);
+		if (dwc->usb2_generic_phy){
+		printk(KERN_INFO "==> %s %d\n", __func__, __LINE__);
+			phy_set_mode(dwc->usb2_generic_phy, PHY_MODE_USB_HOST);
+		}
+		printk(KERN_INFO "==> %s %d\n", __func__, __LINE__);
+
 		ret = dwc3_host_init(dwc);
 		if (ret)
 			dev_err(dwc->dev, "failed to initialize host\n");
 		break;
 	case DWC3_GCTL_PRTCAP_DEVICE:
 		dwc3_event_buffers_setup(dwc);
+
+		printk(KERN_INFO "==> %s %d\n", __func__, __LINE__);
+		if (dwc->usb2_phy)
+			otg_set_vbus(dwc->usb2_phy->otg, false);
+		printk(KERN_INFO "==> %s %d\n", __func__, __LINE__);
+		if (dwc->usb2_generic_phy)
+			phy_set_mode(dwc->usb2_generic_phy, PHY_MODE_USB_DEVICE);
+		printk(KERN_INFO "==> %s %d\n", __func__, __LINE__);
+
 		ret = dwc3_gadget_init(dwc);
 		if (ret)
 			dev_err(dwc->dev, "failed to initialize peripheral\n");
@@ -721,6 +742,8 @@ static void dwc3_core_setup_global_control(struct dwc3 *dwc)
 	dwc3_writel(dwc->regs, DWC3_GCTL, reg);
 }
 
+static int dwc3_core_get_phy(struct dwc3 *dwc);
+
 /**
  * dwc3_core_init - Low-level initialization of DWC3 Core
  * @dwc: Pointer to our controller context structure
@@ -759,6 +782,10 @@ static int dwc3_core_init(struct dwc3 *dwc)
 	if (ret)
 		goto err0;
 
+	ret = dwc3_core_get_phy(dwc);
+	if (ret)
+		goto err0;
+
 	dwc3_core_setup_global_control(dwc);
 	dwc3_core_num_eps(dwc);
 
@@ -903,6 +930,12 @@ static int dwc3_core_init_mode(struct dwc3 *dwc)
 	switch (dwc->dr_mode) {
 	case USB_DR_MODE_PERIPHERAL:
 		dwc3_set_prtcap(dwc, DWC3_GCTL_PRTCAP_DEVICE);
+
+		if (dwc->usb2_phy)
+			otg_set_vbus(dwc->usb2_phy->otg, false);
+		if (dwc->usb2_generic_phy)
+			phy_set_mode(dwc->usb2_generic_phy, PHY_MODE_USB_DEVICE);
+
 		ret = dwc3_gadget_init(dwc);
 		if (ret) {
 			if (ret != -EPROBE_DEFER)
@@ -912,6 +945,12 @@ static int dwc3_core_init_mode(struct dwc3 *dwc)
 		break;
 	case USB_DR_MODE_HOST:
 		dwc3_set_prtcap(dwc, DWC3_GCTL_PRTCAP_HOST);
+
+		if (dwc->usb2_phy)
+			otg_set_vbus(dwc->usb2_phy->otg, true);
+		if (dwc->usb2_generic_phy)
+			phy_set_mode(dwc->usb2_generic_phy, PHY_MODE_USB_HOST);
+
 		ret = dwc3_host_init(dwc);
 		if (ret) {
 			if (ret != -EPROBE_DEFER)
@@ -1148,10 +1187,6 @@ static int dwc3_probe(struct platform_device *pdev)
 	platform_set_drvdata(pdev, dwc);
 	dwc3_cache_hwparams(dwc);
 
-	ret = dwc3_core_get_phy(dwc);
-	if (ret)
-		goto err0;
-
 	spin_lock_init(&dwc->lock);
 
 	pm_runtime_set_active(dev);
diff --git a/drivers/usb/dwc3/dwc3-pci.c b/drivers/usb/dwc3/dwc3-pci.c
index 84a2cebfc712..ca99f430a8b6 100644
--- a/drivers/usb/dwc3/dwc3-pci.c
+++ b/drivers/usb/dwc3/dwc3-pci.c
@@ -109,7 +109,7 @@ static int dwc3_pci_quirks(struct dwc3_pci *dwc)
 		int ret;
 
 		struct property_entry properties[] = {
-			PROPERTY_ENTRY_STRING("dr_mode", "peripheral"),
+//			PROPERTY_ENTRY_STRING("dr_mode", "peripheral"),
 			PROPERTY_ENTRY_BOOL("linux,sysdev_is_parent"),
 			{ }
 		};
diff --git a/drivers/usb/dwc3/ulpi.c b/drivers/usb/dwc3/ulpi.c
index bd86f84f3790..37c096568861 100644
--- a/drivers/usb/dwc3/ulpi.c
+++ b/drivers/usb/dwc3/ulpi.c
@@ -22,7 +22,7 @@
 
 static int dwc3_ulpi_busyloop(struct dwc3 *dwc)
 {
-	unsigned count = 1000;
+	unsigned count = 100000;
 	u32 reg;
 
 	while (count--) {
@@ -41,15 +41,19 @@ static int dwc3_ulpi_read(struct device *dev, u8 addr)
 	u32 reg;
 	int ret;
 
+	printk(KERN_INFO "==> %s %d\n", __func__, __LINE__);
 	reg = DWC3_GUSB2PHYACC_NEWREGREQ | DWC3_ULPI_ADDR(addr);
 	dwc3_writel(dwc->regs, DWC3_GUSB2PHYACC(0), reg);
 
+	printk(KERN_INFO "==> %s %d\n", __func__, __LINE__);
 	ret = dwc3_ulpi_busyloop(dwc);
 	if (ret)
 		return ret;
+	printk(KERN_INFO "==> %s %d\n", __func__, __LINE__);
 
 	reg = dwc3_readl(dwc->regs, DWC3_GUSB2PHYACC(0));
 
+	printk(KERN_INFO "==> %s %d\n", __func__, __LINE__);
 	return DWC3_GUSB2PHYACC_DATA(reg);
 }
 

```